### PR TITLE
feat(commands): add `remove` command

### PR DESCRIPTION
**Pull request by jdrueckert**
Replaces and closes #1 

Adding a command to remove items from inventory.

# How to test
1. Give a few pickaxes using "give pickaxe 'amount' " command
1. Remove one pickaxe using "remove pickaxe" 
1. Remove many pickaxes using "remove pickaxe amount" with various values of the amount
1. Try the remove command when you don't have a pickaxe in the inventory
1. Try the remove command with amount parameter greater than the amount of pickaxe you have in your inventory.

Test the steps 1 to 5 for blocks, shapes and stackable objects, too.

# Outstanding before merging this PR
- [x] removing multiple stackable items
- [x] removing more blocks than are in the inventory (remove those that exist and print how many were removed)

# Follow-up tasks to merging this PR
- overhauling the remove command code (remove code duplication due to sequential handling of item and block removal)
- overhauling the give command code (remove code duplication due to sequential handling of item and block removal)

Co-authored-by: Darshan <darshantank3@gmail.com>

### CODE REVIEW

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 1]
```diff
@@ -33,15 +35,29 @@
...
 
+import java.util.ArrayList;
 import java.util.Collections;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Optional;
 import java.util.Set;
 
 @RegisterSystem
+@Share(ItemCommands.class)
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java:56]: Why is this necessary? If in doubt, add a todo to track this question right here in the code.

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 2]
```diff
@@ -96,16 +114,10 @@ public String give(
...
 
         } else if (matches.size() > 1) {
-            StringBuilder builder = new StringBuilder();
-            builder.append("Requested item \"");
-            builder.append(itemPrefabName);
-            builder.append("\": matches ");
-            Joiner.on(" and ").appendTo(builder, matches);
-            builder.append(". Please fully specify one.");
-            return builder.toString();
+            return buildAmbiguousobjectIdString("item", itemPrefabName, matches);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java:117]: Type in camel-case wording:
```suggestion
            return buildAmbiguousObjectIdString("item", itemPrefabName, matches);
```

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 3]
```diff
@@ -114,6 +126,43 @@ public String give(
...
+            @CommandParam(value = "blockShapeName", required = false) String shapeUriParam) {
+
+        int removalQuantity = quantity != null ? quantity : 1;
+        if (removalQuantity < 1) {
+            return "Invalid quantity of items!";
+        }
+
+        Set<ResourceUrn> matches = assetManager.resolve(inventoryObjectId, Prefab.class);
+
+        if (matches.size() == 1) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java:147]:
```suggestion
        String message;
        if (matches.size() == 1) {
```

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 4]
```diff
@@ -114,6 +126,43 @@ public String give(
...
+            Prefab prefab = assetManager.getAsset(matches.iterator().next(), Prefab.class).orElse(null);
+            String message = removeItem(prefab, quantity, client);
+            if (message != null) {
+                return message;
+            }
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java:152]: I don't think we need to do the `null` check and return early. When defining `String message` before the if-else blocks, this should also work:

```suggestion
            message = removeItem(prefab, quantity, client);
```

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 5]
```diff
@@ -114,6 +126,43 @@ public String give(
...
+        } else if (matches.size() > 1) {
+            return buildAmbiguousobjectIdString("item", inventoryObjectId, matches);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java:154]:
```suggestion
            message = buildAmbiguousObjectIdString("item", inventoryObjectId, matches);
```

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 6]
```diff
@@ -114,6 +126,43 @@ public String give(
...
+        } else {
+            // If no matches are found for items, try blocks
+            String message = removeBlock(client, inventoryObjectId, quantity, shapeUriParam);
+            if (message != null) {
+                return message;
+            }
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java:160]:
```suggestion
            message = removeBlock(client, inventoryObjectId, quantity, shapeUriParam);
```

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 7]
```diff
@@ -114,6 +126,43 @@ public String give(
...
+        }
+
+        return "Could not find an item or block matching \"" + inventoryObjectId + "\"";
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java:163]:
```suggestion
        if (message != null) {
            return message;
        } else {
            return "Could not find an item or block matching \"" + inventoryObjectId + "\"";
        }
```

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 8]
```diff
@@ -170,4 +219,189 @@ public String bulkGiveItem(
         }
         return result;
     }
+
+    private String buildAmbiguousobjectIdString(String objectType, String objectId, Set<ResourceUrn> possibleMatches) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java:223]:
```suggestion
    private String buildAmbiguousObjectIdString(String objectType, String objectId, Set<ResourceUrn> possibleMatches) {
```

File: src/main/java/org/terasology/logic/inventory/ItemCommands.java [hunk 9]
```diff
@@ -170,4 +219,189 @@ public String bulkGiveItem(
...
+                    + (removedItems > 1 ? removedItems + " blocks of " : "a block of ")
+                    + blockFamily.getDisplayName();
+        } else {    // can also mean that all removal attempts failed
+            return "Nothing to remove, you don't have \""
+                    + blockFamily.getDisplayName()
+                    + "\" in your inventory";
+        }
+    }
+
+    private <T extends Comparable<T>> List<T> sortItems(Iterable<T> items) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/inventory/ItemCommands.java:399]: I don't see a call-site of this method. Is it used?

**Comment by jdrueckert** [comment]
@jellysnake will wait for a review from you until April 12, 10PM _your_ time, otherwise merge the PR without it.
